phoenix-core/src/main/java/org/apache/phoenix/monitoring/MetricUtil.java

```diff
@@ -38,4 +46,22 @@ public static MetricsStopWatch getMetricsStopWatch(boolean isRequestMetricsEnabl
         return new MetricsStopWatch(true);
     }
 
+    // We need to cover the case when JmxCacheBuster has just stopped the HBase metrics
+    // system, and not accidentally overwrite the DefaultMetricsSystem singleton.
+    // See PHOENIX-6699
+    public static boolean isDefaultMetricsInitialized() {
+        try {
+            MetricsSystemImpl metrics = (MetricsSystemImpl) DefaultMetricsSystem.instance();
+            Field prefixField = MetricsSystemImpl.class.getDeclaredField("prefix");
```

Comment:
_MetricsSystemImpl.monitoring_ is not useful for this purpose, as it is set to false when the metrics system is stopped.
This is why we can overwrite the stopped Default Metrics System:
https://github.com/apache/hadoop/blob/7bd7725532fd139d2e0e1662df7700f7ab95067a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java#L156

_MetricsSystemImpl.prefix_ is only written by the init() method, it is not touched otherwise, so it can be used to check for the codition when the MetricsSystemImpl is initialized, but in a stopped state:
https://github.com/apache/hadoop/blob/7bd7725532fd139d2e0e1662df7700f7ab95067a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/metrics2/impl/MetricsSystemImpl.java#L154

I don't think we have to worry about a race around prefix, AFAICT by the time the Phoenix classes are loaded, HBase has long initialized DefaultMetrics, and prefix has been set.

Reply:
Of course, this, like many Metrics stuff, is a huge hack, we shouldn't need to use reflection to check if DefaultMetricsSystem has been initialized already.

Also, we could just to see if we are and RS/Master process, and skip the initialization, but I don't know a way to do that.
Maybe you can suggest something ?

Reply:
> I don't think we have to worry about a race around prefix, AFAICT by the time the Phoenix classes are loaded, HBase has long initialized DefaultMetrics, and prefix has been set.

Ah, ok, I wondered if I was completely worrying about a non-issue.

> Of course, this, like many Metrics stuff, is a huge hack, we shouldn't need to use reflection to check if DefaultMetricsSystem has been initialized already.

Yeah, that's life with hadoop metrics2 :)

> Also, we could just to check if we are and RS/Master process, and skip the initialization, but I don't know a way to do that.

Hrm, I think we could do that (let me see what I can dig up). What about Geoffrey's point about the server-side metrics stuff we do have in Phoenix?

Reply:
The server side Phoenix metrics should be added to the Metrics System initialized by HBase, and there should be no change with this patch.
I haven't got around to confirming that on a real cluster yet.

Reply:
> we could just to check if we are and RS/Master process, and skip the initialization, but I don't know a way to do that.

I'm not coming up with any static state that you could look at to try to determine if you're in the context of a RS or Master, shy of _hoping_ that the classpath client side doesnt' have the `HRegionServer.class` in it (or similar).

The other thing I thought of was looking at `RpcServer.getCurrentCall()`. Normally, there is a corresponding user RPC which would result in this returning the current RPC. But, this method would return empty during a compaction (e.g. updating Phoenix stats) where there is no active HBase Call.

Could we somehow separate the Phoenix client-side-only metrics from the execution pathway of the Phoenix server-side metrics? I imagine this is no easier to do.

Reply:
> Could we somehow separate the Phoenix client-side-only metrics from the execution pathway of the Phoenix server-side metrics? I imagine this is no easier to do.

We're veering into territories that are beyond my knowledge of the metrics system.

AFAICT the Hadoop metrics system always uses the DefaultMetrics object, I haven't seen anything that points to using multiple MetricsSystem objects at the same time.

This is mostly fine, as we can add/remove metrics from this default MetricsSystem objects, and this doesn't cause any problems, and this means that we add the additional Phoenix metrics in the RS dynamically, and they can be accessed in a unified way.

Pre HBase 1.4, Phoenix didn't use the metrics2 framework for its client metrics, and it had to be accessed programmatically (no JMX). I don't think we want that.

The only problem comes from the fact that SOMETHING has to initialize a DefaultMetrics object on the Phoenix client side once before the Metrics System can be used, and 
1. HBase (server side) and Phoenix (anywhere) attempt this
2. While DefaultMetrics is protected from re-initialization, the stop/start sequence preformed by JmxCacheBuster defeats this, hence this bug.

I have also tried to simply disable the DefaultMetrics initialization from Phoenix, but in that case the Metrics System wasn't initialized at all on the client side.

All in all, I think that my current solution is good enough, and I'm not convinced that expliring alternatives is worth the time (even though I asked for them from you)

Reply:
> All in all, I think that my current solution is good enough, and I'm not convinced that expliring alternatives is worth the time (even though I asked for them from you)

Yeah, I think your assessment is fair. We carry lots of baggage as a result of Hadoop metrics2. This isn't the worst thing that I've seen done :)